### PR TITLE
v1

### PR DESCRIPTION
**Pull request by michaelwittig**
feat: add new variable to control static monitoring vs anomaly detection
feat: add variables to control period and evaluation periods
breaking: var.xxx_threshold = -1 no longer works, set var.xxx = "off" instead
breaking: var.xxx_threshold = -2 no longer works, set var.xxx = "anomaly_detection" instead

### CODE REVIEW

File: main.tf [hunk 1]
```diff
@@ -25,38 +25,48 @@ data "aws_ec2_instance_type" "instance" {
...
-  ebs_io_credit_balance_evaluation_periods             = min(max(local.ebs_io_credit_balance_evaluation_periods_raw, 1), floor(86400 / local.ebs_io_credit_balance_period))
+  topic_arn                = var.create_topic == false ? var.topic_arn : join("", aws_sns_topic.marbot.*.arn)
+  network                  = lookup(jsondecode(data.http.network.response_body), data.aws_instance.instance.instance_type, {})
+  network_baseline         = lookup(local.network, "baseline", -1)
+  network_burst            = lookup(local.network, "burst", -1)
+  instance_name            = lookup(data.aws_instance.instance.tags, "Name", "")
+  alarm_description_prefix = (local.instance_name == "") ? "" : "${local.instance_name}'s "
+  enabled                  = var.enabled && lookup(data.aws_instance.instance.tags, "marbot", "on") != "off" && lookup(data.aws_instance.instance.tags, "marbot:enabled", "true") != "false"
+
+  cpu_utilization                        = (local.cpu_utilization_threshold < -1.5) ? "anomaly_detection" : ((local.cpu_utilization_threshold < -0.5) ? "off" : lookup(data.aws_instance.instance.tags, "marbot:cpu-utilization", var.cpu_utilization))
```
Comment (andreaswittig) [inline main.tf:36]: Is this for backwards compatibility?

File: main.tf [hunk 2]
```diff
@@ -217,14 +227,25 @@ resource "aws_cloudwatch_metric_alarm" "cpu_utilization_anomaly_detection" {
...
+      expression  = "IF(m1<${local.cpu_utilization_threshold}, ${local.cpu_utilization_threshold}, m1)"
+      label       = "CPUUtilization (threshold)"
+      return_data = "true"
+    }
+  }
+
   metric_query {
     id          = "m1"
-    return_data = "true"
+    return_data = (local.cpu_utilization == "anomaly_detection") ? "true" : null
```
Comment (andreaswittig) [inline main.tf]: Is it possible, that AWS charges for metric queries, that are part of the alarm definition, even if those are not returning data? Maybe use two different resources for the CloudWatch alarm instead?

**Review by andreaswittig — APPROVED** [review]

File: main.tf (hunk 1 above)
Comment (michaelwittig) [reply main.tf:36]: yes. for the tags.

File: main.tf (hunk 2 above)
Comment (michaelwittig) [reply main.tf]: charges should be the same? we read the same metric. the only difference is that we add another IF() depending on anomaly_detection or static_anomaly_detection

Comment (andreaswittig) [reply main.tf]: You are right, it's per metric not per query.

File: main.tf (hunk 1 above)
Comment (andreaswittig) [reply main.tf:36]: 👍

**Pull request merged**